pkg/api/pod/warnings.go

```diff
@@ -243,6 +243,17 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 				}
 			}
 		}
+		// duplicate containers[*].ports
```

Comment:
> * Shouldn't this be ACROSS containers in a pod?

for the client-side apply / strategic merge patch, across containers is not needed?

> * You need to compare with protocol.  It's legit to have the same port number on both TCP and UDP

The code check protocol+hostip+hostport.
https://github.com/kubernetes/kubernetes/blob/36dd5f2846a6003d1b346969183270440f37b190/pkg/apis/core/validation/validation.go#L2786-L2807

> * I think there are people who rely on this - specifically multiple `hostPort` definitions that point to the same `containerPort`

multiple `hostPort` for the same port is valid.

Reply:
For the 3rd question, 

```
[root@paco ~]# cat diffHostPort.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx-1
        image: nginx:1.14.2
        ports:
        - containerPort: 443
          protocol: UDP
          name: b1
          hostPort: 40080
        - containerPort: 443
          protocol: UDP
          name: b2
          hostPort: 30080

[root@paco ~]# kubectl apply --server-side=true -f diffHostPort.yaml
Error from server: failed to create manager for existing fields: failed to convert new object 
(default/nginx; apps/v1, Kind=Deployment) to smd typed:
 .spec.template.spec.containers[name="nginx-1"].ports: duplicate entries for key [containerPort=443,protocol="UDP"]

[root@paco ~]# kubectl apply --server-side=false -f diffHostPort.yaml
deployment.apps/nginx unchanged
```

It may be weried to users here.

Reply:
> Error from server: failed to create manager for existing fields: failed to convert new object 
(default/nginx; apps/v1, Kind=Deployment) to smd typed:
 .spec.template.spec.containers[name="nginx-1"].ports: duplicate entries for key [containerPort=443,protocol="UDP"]

@apelisse this is fun... what's the expected/recommended recovery path for a user trying to fix an error like that using server-side apply? Do they have to use update or client-side apply to fix the error before they can use server-side apply?

Reply:
If the existing object is invalid, there's no way server-side apply is going to be able to process it, so yeah, I think one would have to fix it with edit/csa/some sort of update/patch before being able to SSA.

Reply:
Uggh.  Yeah, it sure weems like the validation lets things through that are not covered by the set of `+listMapKey` defined.  Is it safe to edit the listMapKeys?  We never could do that with `patchMergeKey` because we didn't know what clients are out there.

Reply:
https://github.com/kubernetes/kubernetes/issues/113482 to track

Reply:
I think the listMapKeys are correct, what's incorrect is the validation, no?

Reply:
no, people use multiple duplicate blocks with different `hostPort` pointing at the same containerPort.  It's weird, but I have seen it in the wild.

Reply:
Yeah, I think I've seen that too now that you mention it. We could possibly use `[name,port,protocol]` as the key then? I need to think of the implications of doing that.